core/cpu.c

```diff
@@ -548,13 +556,107 @@ void hax_panic_log(struct vcpu_t *vcpu)
     hax_error("log_vmoff_err %lx\n", log_vmxoff_err);
 }
 
+vmx_error_t vmxroot_enter(struct per_cpu_data *cpu_data);
+vmx_error_t vmxroot_leave(struct per_cpu_data *cpu_data);
+void restore_host_cr(struct per_cpu_data *cpu_data);
+
+/*
+ * Try to enter VMX-ROOT operations.
+ * Must be smp_call_function() safe.
+ * The success/failure logic is done in the caller.
+ */
+vmx_error_t vmxroot_enter(struct per_cpu_data *cpu_data)
+{
+    uint64 fc_msr;
+    uint64 cr0, cr0_f0, cr0_f1;
+    uint64 cr4, cr4_f0, cr4_f1;
+
+    smp_mb();
```

Comment:
I'm not really familiar with memory barriers, and I'm curious why you use it here. Because of this line, `smp_mb()` is called twice for both `load_vmcs()` and `invept_smpfunc()` after the refactoring, but only once in the original code.

Reply:
It's only compiler stuff optimization to prevent memory reordering operations ... Please refer to https://yarchive.net/comp/linux/memory_barriers.html, if you want to know more about it. You don't have to see it as a real function call.
In this situation we have to be safe toward smp function calls (like it's the case in HAXM code when you meet xxx_smpfunc()). Even if the 2 functions use to be called along side, if one day someone calls them appart ... it may fail.
As for all of your previous comments, i tried to keep the philosophy of the existing code without rewriting everything actually.